### ISSUE TITLE
Lua filter buffering while making async HttpCall

### ISSUE THREAD

**Issue by s-vivien** [envoyproxy/envoy#10076]
Hi,

I am using Envoy in the following scenario :

Single listener with a single lua filter.
The lua filter makes a HTTP call to a third party API, extracts some data from the response and modify upstream headers accordingly. This is what it looks like :

```yaml
static_resources:
  listeners:
  - per_connection_buffer_limit_bytes: 1024
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 8080
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stream_idle_timeout: 0s
          codec_type: auto
          stat_prefix: storage-api
          server_name: "storage"
          route_config:
            name: s3_route
            virtual_hosts:
            - name: upload
              domains:
              - "*"
              routes:
              - match:
                  regex: "^.*$"
                  headers:
                   - name: ":method"
                     exact_match: "PUT"
                route:
                  cluster: s3
                  timeout: 0s
          http_filters:
          - name: envoy.lua
            typed_config:
              "@type": type.googleapis.com/envoy.config.filter.http.lua.v2.Lua
              inline_code: |
                function envoy_on_request(request_handle)

                  request_handle:logDebug("- - - - ON REQUEST UPLOAD - - - -")
                  
                  -- third party API call
                  local headers, body = request_handle:httpCall(
                  "thirdparty",
                  {
                   [":method"] = "GET",
                   [":path"] = "/v2/5e480bb3300000e539294a5e?mocky-delay=10s",
                   [":authority"] = "www.mocky.io"
                  },
                  nil,
                  65000)
                  request_handle:logDebug("mocky responded with ")
                  request_handle:logDebug(headers[":status"])
                  request_handle:logDebug(body)
                  

                  -- change upstream header with some information from the async http call
                  request_handle:headers():add("someHeader", body)
                  
                  
                  request_handle:logDebug("- - - - ON REQUEST UPLOAD END - - - -")

                end
          - name: envoy.router
            typed_config: {}
  clusters:
  - name: thirdparty
    connect_timeout: 10.0s
    type: logical_dns 
    dns_lookup_family: V4_ONLY
    lb_policy: round_robin
    load_assignment:
      cluster_name: thirdparty
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: www.mocky.io
                port_value: 80
  - name: s3
    connect_timeout: 10.0s
    type: strict_dns
    hosts:
      socket_address:
        address: 10.0.2.15
        port_value: 9000
  admin:
    access_log_path: "/dev/null"
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 8001

```

Since this API will be used to upload large files, I set the `per_connection_buffer_limit_bytes: 1024` because I don't want Envoy to run OOM.

The problem is that I get a 413 error if the buffer is full before the HTTP call has terminated :

```
[2020-02-16 13:21:24.758][6690][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1932] [C0][S7481487801671065876] request data too large watermark exceeded
[2020-02-16 13:21:24.758][6690][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1167] [C0][S7481487801671065876] Sending local reply with details request_payload_too_large
```

If I increase the buffer limit it works fine, but then the whole file is buffered by Envoy, which I want to avoid.

Is there a way to wait for the HTTP call to finish without buffering anything ? Putting the client on hold and then resuming streaming mode ? Watermarking instead of buffering would be perfect in this case.

Or, is there any alternative without LUA filter to what I'm trying to achieve here ?

EDIT : Subsidiary question about memory management :
When a request, or part of it, is buffered by Envoy, how and when is it cleared from memory ?
From what I've seen, when a request is buffered, it eventually raises the allocated memory size of the process. The process memory remains allocated but if it needs to buffer another request of the same size, it will reuse that allocated memory and won't use more.
i.e. the memory print of the process is equal to the maximum amount of data envoy had to buffer at one point.
Am I right ? If so, is the memory released at some point ? Like with a garbage-collector ?

**Comment by mattklein123** [comment]
Right now the Lua filter always buffered during an HTTP call versus having the option to watermark and apply back pressure. I don't see any reason why we couldn't watermark as well as buffer during an HTTP call (if needed) but this would require some thinking on how to allow the various permutations of what folks want to do. cc @dio

**Comment by s-vivien** [comment]
Interesting. While I'm aware that a proper implementation of this feature would require a design-thinking, how hard would it be to hardcode the desired functioning myself ?

Would returning `Http::FilterDataStatus::StopIterationAndWatermark` instead of `Http::FilterDataStatus::StopIterationAndBuffer` [here](https://github.com/envoyproxy/envoy/blob/5248a4fb7d4c2a3d1fa151f944d3a63f6b7a06cf/source/extensions/filters/http/lua/lua_filter.cc#L67) do the trick ?

**Comment by mattklein123** [comment]
> Would returning Http::FilterDataStatus::StopIterationAndWatermark instead of Http::FilterDataStatus::StopIterationAndBuffer here do the trick ?

Yes I think that will do what you want.

**Comment by s-vivien** [comment]
I can confirm that it works like a charm. Keeping the issue open for a stable/legit implementation of the feature.
Thanks.

**Comment by s-vivien** [comment]
In fact, that simple modification has a weird side-effect that I'm trying to understand.
If I use handle:respond(...) in my LUA script, 50% of the time the client connection is cut and the client gets no response at all. It never happens with the unmodified code. It looks like a race condition problem but I haven't yet figured it out ...

**Comment by s-vivien** [comment]
@mattklein123 : I mentionned on the Discord an issue with the watermarking not triggering a "remote early close" when using handle.respond() in LUA script, thus causing weird client-side behaviors.

I just tried to replace Http::FilterDataStatus::StopIterationAndWatermark with Http::FilterDataStatus::StopIterationNoBuffer and it seems to behave like I want (i.e. triggers a remote early close when I use handle.respond() in the LUA script, and doesn't buffer anything while making HTTP calls).

Does it seems like a legit alternative to you ?
